src/rcc/clocking.rs

```diff
@@ -0,0 +1,292 @@
+    pub always_on: bool,
+    pub auto_start: bool,
+}
+
+impl HighSpeedInternal16RC {
+    /// Applies the selection options to the configuration registers and turns the clock on
+    pub fn configure(&self, rcc: &rcc::RegisterBlock) -> (u32, u8) {
+        rcc.cr.modify(|_, w| {
+            w.hsion()
+                .set_bit()
+                .hsikeron()
+                .bit(self.always_on)
+                .hsiasfs()
+                .bit(self.auto_start)
+        });
+        while rcc.cr.read().hsirdy().bit_is_clear() {}
+        (16_000_000, 0b01)
+    }
+}
+
+/// Medium-speed internal 100 kHz - 48 MHz RC
+#[derive(Clone, Copy)]
+pub struct MediumSpeedInternalRC {
+    freq: u32,
+    auto_cal: bool,
+}
+
+impl MediumSpeedInternalRC {
+    pub fn new(freq: u32, auto_cal: bool) -> Self {
+        MediumSpeedInternalRC { freq, auto_cal }
+    }
+
+    pub fn bits(&self) -> u8 {
+        match self.freq {
+            100_000 => 0b0000,
+            200_000 => 0b0001,
+            400_000 => 0b0010,
+            800_000 => 0b0011,
+            1_000_000 => 0b0100,
+            2_000_000 => 0b0101,
+            4_000_000 => 0b0110,
+            8_000_000 => 0b0111,
+            16_000_000 => 0b1000,
+            24_000_000 => 0b1001,
+            32_000_000 => 0b1010,
+            48_000_000 => 0b1011,
+            _ => panic!("bad MSI speed value!"),
+        }
+    }
+
+    /// Configures the MSI to the specified frequency, and enables hardware
+    /// auto-calibration if requested by enabling (and waiting for) the LSE.
+    pub fn configure(&self, rcc: &rcc::RegisterBlock) -> (u32, u8) {
+        rcc.cr
+            .modify(|_, w| unsafe { w.msirange().bits(self.bits()).msirgsel().set_bit() });
+        while rcc.cr.read().msirdy().bit_is_clear() {}
+
+        if self.auto_cal {
+            // FIXME This... may not work? I'm not sure if I've got a board problem or using
+            // the LSE requires some precondition I'm missing. In either case, LSERDY is never
+            // set by the hardware, so auto_cal doesn't succeed.
+            rcc.apb1enr1.modify(|_, w| w.pwren().set_bit());
+
+            rcc.bdcr.modify(|_, w| w.lseon().clear_bit());
+            while rcc.bdcr.read().lserdy().bit_is_set() {}
+            rcc.bdcr
+                .modify(|_, w| unsafe { w.lsedrv().bits(0b11).lseon().set_bit() });
+            while rcc.bdcr.read().lserdy().bit_is_clear() {}
+            rcc.cr.modify(|_, w| w.msipllen().set_bit());
+        }
+        (self.freq(), 0b00)
+    }
+}
+
+impl InputClock for MediumSpeedInternalRC {
+    fn freq(&self) -> u32 {
+        self.freq
+    }
+}
+
+/// High-speed external 4-48 MHz oscillator
+#[derive(Clone, Copy)]
+pub struct HighSpeedExternalOSC(pub u32);
+impl InputClock for HighSpeedExternalOSC {
+    fn freq(&self) -> u32 {
+        self.0
+    }
+}
+
+impl HighSpeedExternalOSC {
+    /// Turns on the HSE oscillator.
+    ///
+    /// (Should this also configure the pin?)
```

Comment:
I cannot remember right now which GPIO pins are required for HSE, but if you unsure right, you can maybe leave comments here like: "Following GPIO pins must configured as <describe how>: <enumeratino of pins>"

But generally unless it is too complex, I think we should do this in user stead